src/Data/TypedByteArray.hs

```diff
@@ -22,3 +33,19 @@ fromList = TypedByteArray . byteArrayFromList
 {-# INLINE unsafeIndex #-}
 unsafeIndex :: Prim a => TypedByteArray a -> Int -> a
 unsafeIndex (TypedByteArray arr) = indexByteArray arr
+
+-- | Construct a 'TypedByteArray' of the given length by applying the function to each index in @[0..n-1]@.
+generate :: forall a. Prim a => Int -> (Int -> a) -> TypedByteArray a
+generate !n f = runST $ do
+    -- Allocate enough space for n correctly aligned elements
+    arr <- newByteArray $ alignment (undefined :: a) * n
+    intLoop 0 n $ \i -> writeByteArray arr i $ f i
+
+    TypedByteArray <$> unsafeFreezeByteArray arr
+
+intLoop :: Monad m => Int -> Int -> (Int -> m ()) -> m ()
```

Comment:
Since this function is recursive, does GHC manage to specialize the Monad constraint? Or do we need a `SPECIALIZE`, or an `INLINE` in combination with moving the actual loop into an inner closure?